packages/dev-utils/src/devApp/render.tsx

```diff
@@ -0,0 +1,179 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { FC, ComponentType } from 'react';
+import ReactDOM from 'react-dom';
+import { BrowserRouter } from 'react-router-dom';
+import BookmarkIcon from '@material-ui/icons/Bookmark';
+import { ThemeProvider, CssBaseline } from '@material-ui/core';
+import {
+  createApp,
+  SidebarPage,
+  Sidebar,
+  SidebarItem,
+  SidebarSpacer,
+  ApiFactory,
+  createPlugin,
+  ApiTestRegistry,
+  ApiHolder,
+} from '@backstage/core';
+import { lightTheme } from '@backstage/theme';
+import * as defaultApiFactories from './apiFactories';
+
+// TODO(rugvip): export proper plugin type from core that isn't the plugin class
+type BackstagePlugin = ReturnType<typeof createPlugin>;
+
+// DevApp builder that is similar to the App builder API, but creates an App
+// with the purpose of developing one or more plugins inside it.
+class DevAppBuilder {
+  private readonly plugins = new Array<BackstagePlugin>();
+  private readonly factories = new Array<ApiFactory<any, any, any>>();
+
+  // Register one or more plugins to render in the dev app
+  registerPlugin(...plugins: BackstagePlugin[]): DevAppBuilder {
+    this.plugins.push(...plugins);
+    return this;
+  }
+
+  // Register an API factory to add to the app
+  registerApiFactory<Api, Impl, Deps>(
+    factory: ApiFactory<Api, Impl, Deps>,
+  ): DevAppBuilder {
+    this.factories.push(factory);
+    return this;
+  }
+
+  // Build a DevApp component using the resources registered so far
+  build(): ComponentType<{}> {
+    const app = createApp();
+    app.registerApis(this.setupApiRegistry(this.factories));
+    app.registerPlugin(...this.plugins);
+    const AppComponent = app.build();
+
+    const sidebar = this.setupSidebar(this.plugins);
+
+    const DevApp: FC<{}> = () => {
+      return (
+        <ThemeProvider theme={lightTheme}>
+          <CssBaseline>
+            <BrowserRouter>
+              <SidebarPage>
+                {sidebar}
+                <AppComponent />
+              </SidebarPage>
+            </BrowserRouter>
+          </CssBaseline>
+        </ThemeProvider>
+      );
+    };
+
+    return DevApp;
+  }
+
+  // Build and render directory to #root element
+  render(): void {
+    const DevApp = this.build();
+
+    const paths = this.findPluginPaths(this.plugins);
+
+    if (window.location.pathname === '/') {
+      if (!paths.includes('/') && paths.length > 0) {
+        window.location.pathname = paths[0];
+      }
+    }
```

Comment:
maybe wrap the app creation in a `useMemo` and the location change in a `useEffect` to just happen once?

Reply:
This only happens once, it's not a react component render, just a regular method :p